File: bfs_dfs/word_ladder.cpp
// leetcode 127. word ladder
// NEVER USE DFS!!!

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
      unordered_set<string> word_dict(wordList.begin(), wordList.end());
      queue<string> to_visit;
      to_visit.push(beginWord);
      int res = 1;
      while(!to_visit.empty()){
        int len = to_visit.size();
        for(int i = 0; i < len; ++i){
          string word = to_visit.front();
          to_visit.pop();

          if(word == endWord) return res;
          find_ladder(word, word_dict, to_visit);
        }
        res++;
      }
      return 0;
    }

private:
  void find_ladder(string word, unordered_set<string>& dict, queue<string>& to_visit){
    dict.erase(word);
    for (int p = 0; p < (int)word.length(); p++) {
      char letter = word[p];
      for (int k = 0; k < 26; k++) {
        word[p] = 'a' + k;
        if (dict.find(word) != dict.end()) {
          to_visit.push(word);
          dict.erase(word);
        }
      }
      word[p] = letter;
    }
  }
};

File: bfs_dfs/word_ladder.py
# 127. Word Ladder
# medium

# BFS instead of DFS, can use twi way BFS.

class Solution:
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        words = set(wordList)
        if not endWord in words:
            return 0

        steps = 1
        queue = set([beginWord])
        while queue:
            next_queue = set()
            print(len(queue))
            while queue:
                cur = queue.pop()
                if cur == endWord:
                    return steps
                for i in range(len(cur)):
                    for ch in string.ascii_lowercase:
                        trans_word = cur[:i] + ch + cur[i+1:]
                        if trans_word in words:
                            next_queue.add(trans_word)

            for word in next_queue:
                words.remove(word)
            queue = next_queue
            steps += 1
        return 0
